tools/adios2pio-nm/adios2pio-nm-lib.cxx

```diff
@@ -1525,229 +1569,337 @@ int ConvertVariableTimedPutVar(IOVector &bpIO, EngineVector &bpReader,
+        const auto vb_blocks = bpReader[0].BlocksInfo(frameid_var, time_step);
+        for (size_t i = 0; i < vb_blocks.size(); i++)
         {
-            ierr = BP2PIO_ERROR;
-            break;
+            nsteps += vb_blocks[i].Count[0];
         }
-        catch (...)
+
+        /* read frame_id buffer */
+        frame_buffer = (int32_t*)calloc(nsteps, sizeof(int32_t));
+        int tmp_idx = 0;
+        for (size_t i = 0; i < vb_blocks.size(); i++)
         {
-            ierr = BP2PIO_ERROR;
-            break;
+            frameid_var.SetBlockSelection(i);
+            bpReader[0].Get(frameid_var, frame_buffer+tmp_idx, adios2::Mode::Sync);
+            tmp_idx += vb_blocks[i].Count[0];
         }
-    }
-    ERROR_CHECK_RETURN(ierr, err_val, err_cnt, comm)
-
-    MPI_Allreduce(&l_nblocks, &g_nblocks, 1, MPI_INT, MPI_SUM, comm);
 
-    nsteps = g_nblocks / nblocks_per_step;
-    if (g_nblocks != nsteps * nblocks_per_step)
-    {
-        if (debug_out)
-            cout << "rank " << mpirank << ":ERROR in processing darray '" << varname
-                 << "'. Number of blocks = " << g_nblocks
-                 << " does not equal the number of steps * number of writers = "
-                 << nsteps << " * " << nblocks_per_step << " = " << nsteps*nblocks_per_step
-                 << endl;
-    }
+        /* read decomp_id buffer */
+        adios2::Variable<int> decompid_var = bpIO[0].InquireVariable<int>("/__pio__/track/decomp_id/" + variable_name);
+        if (!decompid_var)
+        {
+            return BP2PIO_ERROR;
+        }
+        const auto db_blocks = bpReader[0].BlocksInfo(decompid_var, time_step);
+        decomp_buffer = (int32_t*)calloc(nsteps, sizeof(int32_t));
+        tmp_idx = 0;
+        for (size_t i = 0; i < db_blocks.size(); i++)
+        {
+            decompid_var.SetBlockSelection(i);
+            bpReader[0].Get(decompid_var, decomp_buffer + tmp_idx, adios2::Mode::Sync);
+            tmp_idx += db_blocks[i].Count[0];
+        }
 
-    /* Different decompositions at different frames */
-    char decomp_varname[PIO_MAX_NAME];
-    char frame_varname[PIO_MAX_NAME];
-    char fillval_varname[PIO_MAX_NAME];
-    char decompname[PIO_MAX_NAME];
-    snprintf(decomp_varname, PIO_MAX_NAME, "decomp_id/%s", varname.c_str());
-    snprintf(frame_varname, PIO_MAX_NAME, "frame_id/%s", varname.c_str());
-    snprintf(fillval_varname, PIO_MAX_NAME, "fillval_id/%s", varname.c_str());
-    int decomp_id, frame_id, fillval_exist;
-    char fillval_id[PIO_MAX_NAME];
-
-    // TAHSIN -- THIS IS GETTING CONFUSING. NEED TO THINK ABOUT time steps.
-    for (; ts < nsteps; ++ts)
+        /* read fillval_id buffer */
+        adios2::Variable<T> fillval_var = bpIO[0].InquireVariable<T>("/__pio__/track/fillval_id/" + variable_name);
+        if (fillval_var)
+        {
+            const auto fb_blocks = bpReader[0].BlocksInfo(fillval_var, time_step);
+            fillval_buffer = (char*)calloc(nsteps, sizeof(T));
```

Comment:
Need to handle the calloc() failure case

Reply:
Yes, we should handle the calloc() failure here.